### ISSUE TITLE
Store `depends` as distinct attributes

### ISSUE THREAD

**Issue by djmitche** [GothenburgBitFactory/taskwarrior#2569]
Similar to #2553, and as part of #2507, let's store `depends` as distinct attributes, rather than `,`-separated strings.  This will allow merging of changes to these sets without overwriting.  The change is similar to how annotations are currently stored.

* `depends:uuid1,uuid2` becomes `depends_uuid1: "x", depends_uuid2: "x"`.  The `x` is required since empty strings are regarded as un-set attributes.

**Comment by djmitche** [comment]
@tbabej why would the following behave differently?

```c++
  std::vector <Task> Task::getDependencyTasks () const
  {
    auto depends = get ("depends");
    std::vector <Task> blocking;
    if (depends != "")
      for (auto& it : Context::getContext ().tdb2.pending.get_tasks ())
        if (it.getStatus () != Task::completed &&
            it.getStatus () != Task::deleted   &&   
            depends.find (it.get ("uuid")) != std::string::npos)
          blocking.push_back (it);
  
    return blocking;
  }
```
```c++
  std::vector <Task> Task::getDependencyTasks () const
  {
    std::vector <Task> all;                                                                                                                                                                                                         
    for (auto& dep : split (get ("depends"), ','))
    {
      Task task;
      Context::getContext ().tdb2.get (dep, task);
      all.push_back (task);
    }
  
    return all;
  }
```

With the second, I get
```
not ok 16 - dependencies.t: Check that completing a task unblocks
# FAIL: AssertionError on file /usr/lib/python3.8/unittest/case.py line 60 in testPartExecutor: 'yield':
#       "Unblocked 1 'one'." not found in "Completed task 2 'two'.
# Completed 1 task.
# "
```
It looks like 
```c++
  void feedback_unblocked (const Task& task)
  {
    if (Context::getContext ().verbose ("affected"))
    {
      // Get a list of tasks that depended on this task.
      auto blocked = task.getBlockedTasks (); 
  
      // Scan all the tasks that were blocked by this task
      for (auto& i : blocked)
      {                                                                                                                                                                                                                             
        auto blocking = i.getDependencyTasks (); 
        if (blocking.size () == 0)
        {   
          if (i.id)
            std::cout << format ("Unblocked {1} '{2}'.",
                                 i.id,
                                 i.get ("description"))
                      << "\n";
          else
          {   
            std::string uuid = i.get ("uuid");
            std::cout << format ("Unblocked {1} '{2}'.",
                                 i.get ("uuid"),
                                 i.get ("description"))
                      << "\n";
          }   
        }   
      }   
    }
  }
```
depends on some functions returning data from before the modification, but I can't quite make out which.

**Comment by tbabej** [comment]
Unclear so far from just reading it. I'm gonna test it next now, let me just point out in the meantime that the new version will be a bit less efficient. It is not immediately clear, but `TDB2::get` calls `TF2::get`, which is implemented as:

```c++
bool TF2::get (const std::string& uuid, Task& task)
{
  if (! _loaded_tasks)
    load_tasks ();

  if (_tasks_map.size () > 0 && uuid.size () == 36)
  {
    // Fast lookup, same result as below. Only used during "task import".
    auto i = _tasks_map.find (uuid);
    if (i != _tasks_map.end ())
    {
      task = i->second;
      return true;
    }
  }
  else
  {
    // Slow lookup, same result as above.
    for (auto& i : _tasks)
    {
      if (closeEnough (i.get ("uuid"), uuid, uuid.length ()))
      {
        task = i;
        return true;
      }
    }
  }

  return false;
}
```

in particular, the `TF2::get` is not a dictionary lookup (unless we're dealing with the `import` command). In other words, the `tdb2.get()` call performs a linear search over all tasks, and as such as `O(N)` complexity. This means that the new version of the implementation has `O(M*N)` complexity, where `M` is the number of uuids in the `depends` attribute (compared to `O(N)` before).

**Comment by tbabej** [comment]
Oh, I see it now. You're missing the checks for the status of the blocking tasks. The dependency is not blocking the "blocked" task if the dependency's status is completed or deleted.

**Comment by djmitche** [comment]
Ah, good point about the efficiency.  That's surprising!

And, thanks -- that does explain it!  I had noted that the second choice would return more tasks, but missed how returning more tasks would cause this test to fail.

**Comment by tbabej** [comment]
> Ah, good point about the efficiency. That's surprising!

Yes, that's one of the useful tidbits to keep in mind about the codebase (since the linear scans are common, and intuitively one feels like they should be replaced with more efficient lookups). One of the things on my list to explore is if using a map universally under the hood (like is currently used for `task import`) would actually speed things up.

> And, thanks -- that does explain it! I had noted that the second choice would return more tasks, but missed how returning more tasks would cause this test to fail.

Can't blame you :slightly_smiling_face: the method is slightly misleadingly called `getDependencyTasks` and not `getBlockingTasks`.

**Comment by tbabej** [comment]
Fixed by #2573.
